# 1. Write a function that returns the largest number from a list without using max().

In [ ]:
def largest_number(lst):
    largest = lst[0]
    for i in range(len(lst)):
        if lst[i] > largest:
            largest = lst[i]
    return largest

In [ ]:
print(largest_number([10, 25, 7, 99, 45]))

99


# 2. Write a function that checks whether a given string is a palindrome.

In [ ]:
def is_palindrome(text):
    return text == text[::-1]

print(is_palindrome("madam"))

True


# 3. Write a function that returns character frequencies in a dictionary

In [ ]:
def char_frequency(text):
    freq = {}
    for ch in text:
        if ch in freq:
            freq[ch] += 1
        else:
            freq[ch] = 1
    return freq

In [ ]:
print(char_frequency("python"))

{'p': 1, 'y': 1, 't': 1, 'h': 1, 'o': 1, 'n': 1}


# 4. Find Second Highest Number

In [ ]:
def Second_Highest_Number(lst):
   largest = lst[0]
   second = lst[0]
   for num in lst:
        if num > largest:
            second = largest
            largest = num
        elif num > second and num != largest:
            second = num
            return second

In [ ]:
print(Second_Highest_Number([10, 25, 7, 99, 45]))

45


# 5. Find the Separate Even and Odd Numbers : Return two lists.

In [ ]:
def separate_even_odd(lst):
    even = []
    odd = []
    for num in lst:
        if num % 2 == 0:
            even.append(num)
        else:
            odd.append(num)
    return even, odd

In [ ]:
print(separate_even_odd([1, 2, 3, 4, 5, 6]))

([2, 4, 6], [1, 3, 5])


# 6. Find Duplicate Elements : ls = [1,2,3,2,4,5,1]

In [ ]:
def find_duplicates(lst):
    duplicate = []
    for num in lst:
        if lst.count(num) > 1 and num not in duplicate:
            duplicate.append(num)
    return duplicate

In [ ]:
print(find_duplicates([1, 2, 3, 2, 4, 5, 1]))

[1, 2]


# 7. Count the Words in Sentence input: ("Python is easy to learn")

In [ ]:
def word_count(sentence):
    words = sentence.split()
    count = 0
    for i in range(len(words)):
        count = count+ 1
    return count

In [ ]:
print(word_count("Python is easy to learn"))

5


# 8. Check Student Pass or Fail: Return "Pass" if marks ≥ 35.

In [ ]:
def result(marks):
    if marks >= 35:
        return "Pass"
    else:
        return "Fail"

In [ ]:
print(result(40))

Pass


# 9. Find Common Elements: input two lists : ([1,2,3,4],[3,4,5,6])

In [ ]:
def common_elements(list1, list2):
    common = []
    for item in list1:
        if item in list2:
            common.append(item)
    return common

In [ ]:
print(common_elements([1, 2, 3, 4], [3, 4, 5, 6]))

[3, 4]


# 10. Reverse Each Word (not string): input: I love python

In [ ]:
def reverse_words(sentence):
    words = sentence.split()
    output = ""
    for i in range(len(words)):
        output = output + words[i][::-1] + " "
    return output.strip()

In [ ]:
print(reverse_words("I love python"))

I evol nohtyp


# 11. Find employee with highest salary.
input: employees = {
    "John":50000,
    "Priya":70000,
    "David":60000
}


In [ ]:
def highest_salary(emp):
    name = ""
    salary = 0
    for i in emp:
        if emp[i] > salary:
            salary = emp[i]
            name = i
    return name, salary
employees = {
    "John": 50000,
    "Priya": 70000,
    "David": 60000
}

In [ ]:
print(highest_salary(employees))

('Priya', 70000)


# 12. Remove Duplicates Without Set : input: [1,2,3,2,4,1,5]


In [ ]:
def remove_duplicates(lst):
    unique = []
    for item in lst:
        if item not in unique:
            unique.append(item)
    return unique

In [ ]:
print(remove_duplicates([1, 2, 3, 2, 4, 1, 5]))

[1, 2, 3, 4, 5]


# 13. Count Uppercase, Lowercase, Digits: input : "PyThOn123")

In [1]:
def count_characters(text):
    upper = 0
    lower = 0
    digits = 0
    for ch in text:
        if ch.isupper():
            upper =upper+ 1
        elif ch.islower():
            lower =lower+ 1
        elif ch.isdigit():
            digits =digits+ 1
    return upper, lower, digits
count_characters("PyThOn123")

(3, 3, 3)

# 14. Product Sales Analysis (Real-Time Data Analyst Question) : Find total sales and highest sale.

In [2]:
def sales_analysis(sales):
    total_sales = sum(sales)
    highest_sale = max(sales)
    return total_sales, highest_sale
sales = [1000,2000,5000,3000]
sales_analysis(sales)

(11000, 5000)

#15. Create a function that returns:
input: sales = {
    "Jan":[100,200,300],
    "Feb":[400,500,100],
    "Mar":[300,600,200]
}
- output:
{
    "Jan":600,
    "Feb":1000,
    "Mar":1100
}


In [3]:
def monthly_totals(data):
    result = {}
    for month in data:
        result[month] = sum(data[month])
    return result
sales = {
    "Jan":[100,200,300],
    "Feb":[400,500,100],
    "Mar":[300,600,200]
}
monthly_totals(sales)

{'Jan': 600, 'Feb': 1000, 'Mar': 1100}

# 16. Employee with Highest Average Rating
ratings = {
    "John":[4,5,3],
    "Priya":[5,5,4],
    "David":[3,4,4]
}
- Output: Priya

In [4]:
def highest_rating_employee(ratings):
    highest_average = 0
    employee = ""
    for name in ratings:
        average = sum(ratings[name]) / len(ratings[name])
        if average > highest_average:
            highest_average = average
            employee = name
    return employee
ratings = {
    "John":[4,5,3],
    "Priya":[5,5,4],
    "David":[3,4,4]
}
highest_rating_employee(ratings)

'Priya'

# 17. Most Frequent Word
text = "python sql python excel sql python"
- Output: python


In [5]:
def most_frequent_word(text):
    words = text.split()
    frequency = {}
    for word in words:
        if word in frequency:
            frequency[word] += 1
        else:
            frequency[word] = 1
    highest = 0
    result = ""
    for word in frequency:
        if frequency[word] > highest:
            highest = frequency[word]
            result = word
    return result
text = "python sql python excel sql python"
most_frequent_word(text)

'python'

# 18. Count Occurrence of Each Number
Input: [1,2,2,3,3,3,4]
= Output:
{
 1:1,
 2:2,
 3:3,
 4:1
}


In [6]:
def count_occurrences(numbers):
    frequency = {}
    for num in numbers:
        if num in frequency:
            frequency[num] += 1
        else:
            frequency[num] = 1
    return frequency
count_occurrences([1,2,2,3,3,3,4])

{1: 1, 2: 2, 3: 3, 4: 1}

# 19. Find Longest Word
Input: "Data Analytics is transforming businesses"
- Output: transforming


In [7]:
def longest_word(sentence):
    words = sentence.split()
    longest = words[0]
    for word in words:
        if len(word) > len(longest):
            longest = word
    return longest
sentence = "Data Analytics is transforming businesses"
longest_word(sentence)

'transforming'

# 20. Create Student Grade Dictionary
Input:
students = [
("John",78),
("Priya",92),
("David",45)
]
- Output:
{
'John':'B',
'Priya':'A',
'David':'F'
}
Rules:
A >= 90
B >= 70
C >= 50
F < 50


In [8]:
def student_grades(students):
    grades = {}
    for name, marks in students:
        if marks >= 90:
            grades[name] = "A"
        elif marks >= 70:
            grades[name] = "B"
        elif marks >= 50:
            grades[name] = "C"
        else:
            grades[name] = "F"
    return grades
students = [
    ("John",78),
    ("Priya",92),
    ("David",45)
]
student_grades(students)

{'John': 'B', 'Priya': 'A', 'David': 'F'}

# 21. Find First Non-Repeating Character
Input: "aabbccdeeff"
- Output: d


In [9]:
def first_non_repeating(text):
    for ch in text:
        if text.count(ch) == 1:
            return ch
first_non_repeating("aabbccdeeff")

'd'

# 22. Monthly Sales Analysis
sales = {
    "Jan":10000,
    "Feb":15000,
    "Mar":12000,
    "Apr":18000
}


In [3]:
sales = {
    "Jan": 10000,
    "Feb": 15000,
    "Mar": 12000,
    "Apr": 18000
}
total_sales = sum(sales.values())
average_sales = total_sales / len(sales)
highest_month = max(sales, key=sales.get)
lowest_month = min(sales, key=sales.get)

print("Total Sales:", total_sales)
print("Average Sales:", average_sales)
print("Highest Sales Month:", highest_month, "-", sales[highest_month])
print("Lowest Sales Month:", lowest_month, "-", sales[lowest_month])

Total Sales: 55000
Average Sales: 13750.0
Highest Sales Month: Apr - 18000
Lowest Sales Month: Jan - 10000


# 22. Create a function that returns:
Total Sales
Average Sales
Highest Month
Expected:
Total = 55000
Average = 13750
Highest = Apr






In [11]:
def monthly_sales_report(sales):
    total = 0
    highest_month = ""
    highest_sale = 0
    for month in sales:
        total = total + sales[month]
        if sales[month] > highest_sale:
            highest_sale = sales[month]
            highest_month = month
    average = total / len(sales)
    return total, average, highest_month
sales = {
    "Jan":10000,
    "Feb":15000,
    "Mar":12000,
    "Apr":18000
}
monthly_sales_report(sales)

(55000, 13750.0, 'Apr')

# 23. Find Customer Purchase Frequency
customers = [
"John",
"Priya",
"John",
"David",
"Priya",
"John"]
- Output:
{
'John':3,
'Priya':2,
'David':1
}

In [12]:
def purchase_frequency(customers):
    frequency = {}
    for customer in customers:
        if customer in frequency:
            frequency[customer] = frequency[customer] + 1
        else:
            frequency[customer] = 1
    return frequency
customers = ["John","Priya","John","David","Priya","John"]
purchase_frequency(customers)

{'John': 3, 'Priya': 2, 'David': 1}

# 24. Find Top 3 Sales
- Input: [1000,5000,7000,2000,9000,8000]
- Output: [9000,8000,7000]

In [13]:
def top_three_sales(numbers):
    numbers.sort()
    return numbers[-3:][::-1]
sales = [1000,5000,7000,2000,9000,8000]
top_three_sales(sales)

[9000, 8000, 7000]

# 25. Product Performance
products = {
"Mobile":25000,
"Laptop":50000,
"Mouse":8000,
"Keyboard":12000
}
Return:
- Best Product : Laptop
- Worst Product : Mouse

In [14]:
def product_performance(products):
    best_product = ""
    worst_product = ""
    highest_sale = 0
    lowest_sale = 999999
    for product in products:
        if products[product] > highest_sale:
            highest_sale = products[product]
            best_product = product
        if products[product] < lowest_sale:
            lowest_sale = products[product]
            worst_product = product
    return best_product, worst_product
products = {
    "Mobile":25000,
    "Laptop":50000,
    "Mouse":8000,
    "Keyboard":12000
}
print("Best product:",product_performance(products)[0])
print("Worst product:",product_performance(products)[1])
product_performance(products)

Best product: Laptop
Worst product: Mouse


('Laptop', 'Mouse')

# 26. Revenue by Category
orders = [
("Mobile",10000),
("Laptop",50000),
("Mobile",15000),
("Mouse",5000)
]
- Output:
{
'Mobile':25000,
'Laptop':50000,
'Mouse':5000
}


In [15]:
def revenue_by_category(orders):
    revenue = {}
    for product, amount in orders:
        if product in revenue:
            revenue[product] = revenue[product] + amount
        else:
            revenue[product] = amount
    return revenue
orders = [
("Mobile",10000),
("Laptop",50000),
("Mobile",15000),
("Mouse",5000)
]
revenue_by_category(orders)

{'Mobile': 25000, 'Laptop': 50000, 'Mouse': 5000}

# 27. Find Pairs Whose Sum = Target
- Input:
nums = [2,7,11,15,3,6]
target = 9
- Output: [(2,7),(3,6)]


In [16]:
def pair_sum(numbers, target):
    pairs = []
    for i in range(len(numbers)):
        for j in range(i + 1, len(numbers)):
            if numbers[i] + numbers[j] == target:
                pairs.append((numbers[i], numbers[j]))
    return pairs
pair_sum([2,7,11,15,3,6], 9)

[(2, 7), (3, 6)]

# 28. Compress String
- Input: aaabbccdddd
- Output: a3b2c2d4


In [17]:
def compress(text):
    result = ""
    count = 1
    for i in range(1, len(text)):
        if text[i] == text[i-1]:
            count = count + 1
        else:
            result = result + text[i-1] + str(count)
            count = 1
    result = result + text[-1] + str(count)
    return result
compress("aaabbccdddd")

'a3b2c2d4'

# 29. Expand String
- Input: a3b2c2
- Output: aaabbcc


In [18]:
def expand(text):
    result = ""
    for i in range(0, len(text), 2):
        character = text[i]
        count = int(text[i+1])
        result = result + character * count
    return result
expand("a3b2c2")

'aaabbcc'

# 30. Find Consecutive Duplicate Count
- Input: [1,1,1,2,2,3,3,3,3]
- Output:
{
1:3,
2:2,
3:4}

In [19]:
def consecutive_count(numbers):
    result = {}
    current = numbers[0]
    count = 1
    for i in range(1, len(numbers)):
        if numbers[i] == current:
            count = count + 1
        else:
            result[current] = count
            current = numbers[i]
            count = 1
    result[current] = count
    return result
numbers = [1,1,1,2,2,3,3,3,3]
consecutive_count(numbers)

{1: 3, 2: 2, 3: 4}

# 31. Nested Dictionary Analysis
sales = {
    "North":{
        "Jan":1000,
        "Feb":2000
    },
    "South":{
        "Jan":3000,
        "Feb":1500
    }
}
- Output:
North Total = 3000
South Total = 4500
Grand Total = 7500


In [20]:
def nested_sales_analysis(sales):
    grand_total = 0
    for region in sales:
        region_total = 0
        for month in sales[region]:
            region_total = region_total + sales[region][month]
        print(region, "Total =", region_total)
        grand_total = grand_total + region_total
    print("Grand Total =", grand_total)
sales = {
    "North":{
        "Jan":1000,
        "Feb":2000
    },
    "South":{
        "Jan":3000,
        "Feb":1500
    }
}
nested_sales_analysis(sales)

North Total = 3000
South Total = 4500
Grand Total = 7500


# 32. Customer Lifetime Value
customers = {
"John":[1000,2000,1500],
"Priya":[5000,3000],
"David":[1000]
}
- Output:
{
'John':4500,
'Priya':8000,
'David':1000
}
- Find:
•	Total spend per customer
•	Highest-value customer
•	Average customer spend



In [21]:
def customer_lifetime_value(customers):
    totals = {}
    highest_customer = ""
    highest_spend = 0
    total_spend = 0
    for customer in customers:
        customer_total = 0
        for amount in customers[customer]:
            customer_total = customer_total + amount
        totals[customer] = customer_total
        total_spend = total_spend + customer_total
        if customer_total > highest_spend:
            highest_spend = customer_total
            highest_customer = customer
    average_spend = total_spend / len(customers)
    print("Customer Totals =", totals)
    print("Highest Customer =", highest_customer)
    print("Average Spend =", average_spend)
customers = {
    "John":[1000,2000,1500],
    "Priya":[5000,3000],
    "David":[1000]
}
customer_lifetime_value(customers)

Customer Totals = {'John': 4500, 'Priya': 8000, 'David': 1000}
Highest Customer = Priya
Average Spend = 4500.0


# 33. Build a Mini Dashboard Function
Input:
sales = [1000,2000,5000,3000,7000]
Return:
Total Sales
Average Sales
Highest Sales
Lowest Sales
Sales Count
- output:
in a dictionary:
{
'Total':18000,
'Average':3600,
'Highest':7000,
'Lowest':1000,
'Count':5
}


In [22]:
def dashboard(sales):
    total = 0
    highest = sales[0]
    lowest = sales[0]
    count = len(sales)
    for sale in sales:
        total = total + sale
        if sale > highest:
            highest = sale
        if sale < lowest:
            lowest = sale
    average = total / count
    result = {
        "Total": total,
        "Average": average,
        "Highest": highest,
        "Lowest": lowest,
        "Count": count
    }
    return result
sales = [1000,2000,5000,3000,7000]
dashboard(sales)


{'Total': 18000,
 'Average': 3600.0,
 'Highest': 7000,
 'Lowest': 1000,
 'Count': 5}

# 34. Validate Password
- Rules:
Minimum 8 chars
At least 1 uppercase
At least 1 lowercase
At least 1 digit
Return: Valid Or Invalid


In [23]:
def validate_password(password):
    if len(password) < 8:
        return "Invalid"
    has_upper = False
    has_lower = False
    has_digit = False
    for ch in password:
        if ch.isupper():
            has_upper = True
        elif ch.islower():
            has_lower = True
        elif ch.isdigit():
            has_digit = True
    if has_upper and has_lower and has_digit:
        return "Valid"
    return "Invalid"
validate_password("Python123")

'Valid'

**# USE arge & kwargs:**

# 1. Create a function that accepts any number of sales values and returns the total sales.



In [24]:
def total_sales(*sales):
    total = 0
    for sale in sales:
        total = total + sale
    return total
total_sales(1000,2000,3000,4000)

10000

#2. Accept any number of sales values and return the highest.



In [25]:
def highest_sale(*sales):
    highest = sales[0]
    for sale in sales:
        if sale > highest:
            highest = sale
    return highest
highest_sale(1000,5000,3000,7000)

7000

# 3. Count Vowels From Multiple Words
- Input: count_vowels("python","analytics","excel")


In [26]:
def count_vowels(*words):
    count = 0
    vowels = "aeiouAEIOU"
    for word in words:
        for letter in word:
            if letter in vowels:
                count = count + 1
    return count
count_vowels("python","analytics","excel")

6

 # 4. Create a functiona and calculate Average of Unlimited Marks

In [27]:
def average_marks(*marks):
    total = 0
    for mark in marks:
        total = total + mark
    average = total / len(marks)
    return average
average_marks(80,90,70,60)

75.0

# 5. Employee Salary Analysis
Accept employee salaries using **kwargs.
- Return:
•	Total Salary
•	Average Salary
- input:
salary_analysis(
        John=50000,
        Priya=60000,
        David=55000

In [28]:
def salary_analysis(**salaries):
    total_salary = 0
    for employee in salaries:
        total_salary = total_salary + salaries[employee]
    average_salary = total_salary / len(salaries)
    print("Total Salary =", total_salary)
    print("Average Salary =", average_salary)
salary_analysis(
    John=50000,
    Priya=60000,
    David=55000
)

Total Salary = 165000
Average Salary = 55000.0


# 6. Create Student Result Card
- Input:
student_result(
John=78,
Priya=95,
David=45
)
- Output:
{
'John':'Pass',
'Priya':'Pass',
'David':'Fail'
}

In [2]:
def student_result(**students):
    result = {}
    for name in students:
        if students[name] >= 35:
            result[name] = "Pass"
        else:
            result[name] = "Fail"
    return result
print(
    student_result(
        John=78,
        Priya=95,
        David=30
    )
)

{'John': 'Pass', 'Priya': 'Pass', 'David': 'Fail'}


# 7. Combine *args and **kwargs
Calculate total sales and display region information.
- Example:
sales_dashboard(
1000,2000,3000,
region="South",
manager="Priya"
)

In [30]:
def sales_dashboard(*sales, **details):
    total = 0
    for sale in sales:
        total = total + sale
    print("Total Sales =", total)
    for key in details:
        print(key, "=", details[key])
sales_dashboard(
    1000,
    2000,
    3000,
    region="South",
    manager="Priya"
)

Total Sales = 6000
region = South
manager = Priya


# 8. Create mini dashboard:
def dashboard(*sales, **details):

- Input:
dashboard(
1000,
2000,
3000,
5000,
region="South",
manager="Priya",
year=2025
)
- Expected Output:
Total Sales : 11000
Average Sales : 2750
Highest Sale : 5000

Region : South
Manager : Priya
Year : 2025

In [31]:
def dashboard(*sales, **details):
    total = 0
    highest = sales[0]
    for sale in sales:
        total = total + sale
        if sale > highest:
            highest = sale
    average = total / len(sales)
    print("Total Sales :", total)
    print("Average Sales :", average)
    print("Highest Sale :", highest)
    print()
    for key in details:

        print(key.title(), ":", details[key])
dashboard(
    1000,
    2000,
    3000,
    5000,
    region="South",
    manager="Priya",
    year=2025
)

Total Sales : 11000
Average Sales : 2750.0
Highest Sale : 5000

Region : South
Manager : Priya
Year : 2025
